### PR TITLE
optimized the workload store struct

### PR DESCRIPTION
**Pull request by zhlsunshine**
optimized the workload store struct:
1. No `Upstream` contains in `WorkloadStore`, eliminating multiple workload clone for `vip` of `WorkloadStore`.
2. Construct `Upstream` when necessary.

### CODE REVIEW

File: src/workload.rs [hunk 1]
```diff
@@ -438,15 +438,20 @@ impl WorkloadStore {
...
-            let mut us: Upstream = us.clone();
-            Self::set_gateway_address(&mut us, hbone_port);
-            debug!("found upstream from VIP: {}", us);
-            return Some(us);
+            let (workload_ip, target_sock_addr) =
+                wl_vips.iter().choose(&mut rand::thread_rng()).unwrap();
+            if let Some(wl) = self.workloads.get(workload_ip) {
+                let mut us = Upstream {
+                    workload: wl.clone(),
+                    port: target_sock_addr.port(),
```
Comment (howardjohn) [inline src/workload.rs:449]: why do we need to store the whole target_sock_addr when we only use the port? Can the `vips` map just be SocketAddr -> SocketAddr (service:serviceport -> workload:workloadport)?

Comment (zhlsunshine) [reply src/workload.rs:449]: Hi @howardjohn, sure, how about define the `vips` like below?
```
vips: HashMap<SocketAddr, HashSet<(serviceport, IpAddr)>>
```
As you know, we just want to use service port and workload ip to construct the Upstream. Is that okay?

Comment (howardjohn) [reply src/workload.rs:449]: I think you could just turn the tuple into a SocketAddr but yeah basically the same

Comment (KfreeZ) [reply src/workload.rs:449]: can we make sure we are not going to lose any information if use SocketAddr to take place of Workload?

Comment (zhlsunshine) [reply src/workload.rs:449]: @KfreeZ according to the workload definition, it's enough for us to identify a workload.

Comment (KfreeZ) [reply src/workload.rs:449]: > @KfreeZ according to the workload definition, it's enough for us to identify a workload.

suppose if we want to use other information of a workload, let's say waypoint address, we have to look it up again in the workloads Hashmap, correct?

Comment (zhlsunshine) [reply src/workload.rs]: Yes, it is

Comment (KfreeZ) [reply src/workload.rs]: > 

cool, thanks for the response, no more question.

File: src/workload.rs [hunk 2]
```diff
@@ -366,7 +366,9 @@ pub struct WorkloadStore {
     workloads: HashMap<IpAddr, Workload>,
     // workload_to_vip maintains a mapping of workload IP to VIP. This is used only to handle removals.
     workload_to_vip: HashMap<IpAddr, HashSet<(SocketAddr, u16)>>,
-    vips: HashMap<SocketAddr, HashSet<Upstream>>,
+    // vips maintains a mapping of socket address with service port to workload ip and socket address
+    // with target ports in hashset.
+    vips: HashMap<SocketAddr, HashSet<(u16, IpAddr)>>,
```
Comment (hzxuzhonghu) [inline src/workload.rs:371]: to be consistent with workload_to_vip, maybe HashSet<(IpAddr, u16)

Comment (zhlsunshine) [reply src/workload.rs:371]: okay, it makes sense

Comment (zhlsunshine) [reply src/workload.rs]: @hzxuzhonghu done!

**Review by howardjohn — APPROVED** [review]

**Pull request merged**